### ISSUE TITLE
File Descriptor leak on open() when using AMQPS protocol

### ISSUE THREAD

**Issue by mattixtech** [Azure/azure-iot-sdk-java#1135]
# Context
Calling `open()` on the IoT hub device client leaks file descriptors when unable to connect when the AMQPS protocol is used.

- **OS and version used:** macOS 11.1, Ubuntu Core 18, Ubuntu Core 20
- **Java runtime used:** tested w/ Java 1.8 & Java 11
- **SDK version used:** 1.29.2

## Description of the issue
When attempting to `open()` the IoT device client connection when the network is down, the client leaks file descriptors depending on how many times `open()` is called. This seems to be specific to using the AMQPS protocol. This is an issue when attempting reconnect logic. For example if you continue trying to connect while the network is down until you get a successful connection, the number of FDs grows until the process runs out of FDs.

This might happen when connecting for the first time while the network is down or when processing the connection status callback for a non retryable error that requires a new client to be instantiated and `open()`ed.

If/when the client does connect and enters steady state, the leaked FDs appear to remain permanently leaked and the number allocated to the process does not appear to decrease.

I tested this issue using MQTT and could not reproduce the issue, therefore I suspect it may be specific to AMQPS, I did not test any other protocol combinations.

## Code sample exhibiting the issue
Please see reproduction example here: https://github.com/mattixtech/iot-device-client-bug-demo/blob/master/src/main/java/org/example/Example.java

Simply run this program with a valid connection string without a network connection and observe the number of file descriptors continue to grow unbounded. For example on macOS I can run this program and monitor the number of FDs with the following command:
```
watch "lsof -p <pid> | wc -l"
```

## Console log of the issue
N/A

AB#9435166

**Comment by timtay-microsoft** [comment]
The open call to a device client doesn't have retry logic, so can you clarify the scenario a bit more? Is it that a file descriptor is leaked when you call open and the open isn't successful, or is it leaked during reconnection logic on a previously established connection?

**Comment by timtay-microsoft** [comment]
Ah, I just saw your sample code. So the issue here is just when you try to open a device client that previously failed to open. We'll try to look into this soon. Thanks for filing the issue!

**Comment by mattixtech** [comment]
> The open call to a device client doesn't have retry logic, so can you clarify the scenario a bit more? Is it that a file descriptor is leaked when you call open and the open isn't successful, or is it leaked during reconnection logic on a previously established connection?

Ultimately if there is no network connection, the call to `open()` fails with an exception and the leak happens at this point.

Calling `closeNow()` or not seems to make no difference, instantiating a new client also has the same problem since the original instance has alread leaked.

**Comment by mattixtech** [comment]
> Ah, I just saw your sample code. So the issue here is just when you try to open a device client that previously failed to open. We'll try to look into this soon. Thanks for filing the issue!

One more thing to point out, the issue isn't necessarily calling open on a client that previously failed. If you create a new client on each failure it still leaks so the leak seems to happen on the first failed attempt not only on subsequent attempts.

**Comment by timtay-microsoft** [comment]
Yep, looks like the exception handling logic in the AMQP open call does fail to free the proton-j reactor if it encounters an exception, so this is definitely a bug in this SDK. The fix should be checked in soon

**Comment by mattixtech** [comment]
> Yep, looks like the exception handling logic in the AMQP open call does fail to free the proton-j reactor if it encounters an exception, so this is definitely a bug in this SDK. The fix should be checked in soon

Glad to hear you have identified the cause, looking forward to being able to try out the fix. Thanks!

**Comment by mattixtech** [comment]
Thanks @timtay-microsoft for looking at this so quick! I hope to get a chance to build the fix and give it a spin soon. Also not sure if you saw but I had a couple comments in the PR regarding a potential other case where this may occur (maybe it is a non-issue): https://github.com/Azure/azure-iot-sdk-java/pull/1136#issuecomment-789157967

**Comment by timtay-microsoft** [comment]
@mattixtech good point. I'll go ahead and open another PR for that fix as well. I do see that the reactor isn't always free'd there, too. #1136 should ensure that the close logic behaves similarly to the open logic

**Comment by mattixtech** [comment]
@timtay-microsoft thanks a lot for the quick turnaround on this issue. What is the timeline like in for getting a new release of the device client published that we can depend on that contains the fixes?

**Comment by timtay-microsoft** [comment]
I'll try to get a release with this fix out before the end of this week. Would that work for you?

**Comment by mattixtech** [comment]
> I'll try to get a release with this fix out before the end of this week. Would that work for you?

Yes, thanks!!

**Comment by timtay-microsoft** [comment]
Closing this issue as this bug has been fixed as of Device Client version 1.29.3. Feel free to open another issue if you still see any other leaks

**Comment by mattixtech** [comment]
Hi @timtay-microsoft, I just tested out 1.29.3 and I'm still running into the same issue. Appears that it is still occurring in both the example code I provided as well as our application. Should I open a new issue?

**Comment by timtay-microsoft** [comment]
My apologies, there was a small mistake I made with the previous fix. #1147 should fix the leaked resources from a failed open call part of this issue. The leaked threads from a failed close call should be fixed by the last release though. Can you confirm that you are seeing at least the latter working as expected?

**Comment by mattixtech** [comment]
> My apologies, there was a small mistake I made with the previous fix. #1147 should fix the leaked resources from a failed open call part of this issue. The leaked threads from a failed close call should be fixed by the last release though. Can you confirm that you are seeing at least the latter working as expected?

In the case of opening and closing a connection without a failure it looks like the resources are released.

In terms of testing the case where a `close()` call failed, I don't think I have an easy way to verify that.

**Comment by timtay-microsoft** [comment]
> In terms of testing the case where a `close()` call failed, I don't think I have an easy way to verify that.

Gotcha, that would be tricky to invoke. Feel free to build from source on the branch for #1147 to test out the failure case then. From my initial testing, it works.

**Comment by timtay-microsoft** [comment]
Device Client version 1.29.4 has been released with the fix for this issue. Can you confirm that you aren't seeing any leaked file descriptors on your end after upgrading?

**Comment by mattixtech** [comment]
> Device Client version 1.29.4 has been released with the fix for this issue. Can you confirm that you aren't seeing any leaked file descriptors on your end after upgrading?

Hey, thanks a lot for releasing a new version. I wasn't able to get around to testing it yet but I will try to report back ASAP early next week.

**Comment by timtay-microsoft** [comment]
Can you share the exception stacktrace that you see when you run your example code? When I disable my internet and run that code, I see something like 

```
com.microsoft.azure.sdk.iot.device.transport.amqps.exceptions.ProtonIOException
	at com.microsoft.azure.sdk.iot.device.transport.amqps.AmqpsExceptionTranslator.convertFromAmqpException(AmqpsExceptionTranslator.java:84)
	at com.microsoft.azure.sdk.iot.device.transport.amqps.AmqpsIotHubConnection.onTransportError(AmqpsIotHubConnection.java:506)
	at org.apache.qpid.proton.engine.BaseHandler.handle(BaseHandler.java:191)
	at org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:108)
	at org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:324)
	at org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:291)
	at com.microsoft.azure.sdk.iot.device.transport.amqps.IotHubReactor.run(IotHubReactor.java:26)
	at com.microsoft.azure.sdk.iot.device.transport.amqps.AmqpsIotHubConnection$ReactorRunner.call(AmqpsIotHubConnection.java:1142)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by mattixtech** [comment]
> Can you share the exception stacktrace that you see when you run your example code? When I disable my internet and run that code, I see something like
> 
> ```
> com.microsoft.azure.sdk.iot.device.transport.amqps.exceptions.ProtonIOException
> 	at com.microsoft.azure.sdk.iot.device.transport.amqps.AmqpsExceptionTranslator.convertFromAmqpException(AmqpsExceptionTranslator.java:84)
> 	at com.microsoft.azure.sdk.iot.device.transport.amqps.AmqpsIotHubConnection.onTransportError(AmqpsIotHubConnection.java:506)
> 	at org.apache.qpid.proton.engine.BaseHandler.handle(BaseHandler.java:191)
> 	at org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:108)
> 	at org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:324)
> 	at org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:291)
> 	at com.microsoft.azure.sdk.iot.device.transport.amqps.IotHubReactor.run(IotHubReactor.java:26)
> 	at com.microsoft.azure.sdk.iot.device.transport.amqps.AmqpsIotHubConnection$ReactorRunner.call(AmqpsIotHubConnection.java:1142)
> 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
> 	at java.lang.Thread.run(Thread.java:748)
> ```
> 
> When I see this error, the AMQP layer is successfully free'ing the reactor, so I'm curious if you are encountering a different exception and are hitting a different code path

```
java.io.IOException: Could not open the connection
	at com.microsoft.azure.sdk.iot.device.DeviceIO.open(DeviceIO.java:162)
	at com.microsoft.azure.sdk.iot.device.InternalClient.open(InternalClient.java:176)
	at com.microsoft.azure.sdk.iot.device.DeviceClient.open(DeviceClient.java:438)
	at org.example.Example.main(Example.java:18)
Caused by: com.microsoft.azure.sdk.iot.device.transport.amqps.exceptions.ProtonIOException
	at com.microsoft.azure.sdk.iot.device.transport.amqps.AmqpsExceptionTranslator.convertFromAmqpException(AmqpsExceptionTranslator.java:84)
	at com.microsoft.azure.sdk.iot.device.transport.amqps.AmqpsIotHubConnection.onTransportError(AmqpsIotHubConnection.java:506)
	at org.apache.qpid.proton.engine.BaseHandler.handle(BaseHandler.java:191)
	at org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:108)
	at org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:324)
	at org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:291)
	at com.microsoft.azure.sdk.iot.device.transport.amqps.IotHubReactor.run(IotHubReactor.java:26)
	at com.microsoft.azure.sdk.iot.device.transport.amqps.AmqpsIotHubConnection$ReactorRunner.call(AmqpsIotHubConnection.java:1142)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by timtay-microsoft** [comment]
After digging into this a bit further, I can confirm that the issue is still present, yes, and it impacts windows and linux as well as MacOS. I can also confirm that the IoT Hub service client also has this same issue. I'm still digging into why freeing the reactor isn't sufficient for releasing these resources though

**Comment by timtay-microsoft** [comment]
After even further digging, this is beginning to look like a bug in our AMQP library proton-j. I can't write even a basic sample amqp connection that handles this scenario without leaking file descriptors. I went ahead and filed a bug on the proton-j library itself [here](https://issues.apache.org/jira/browse/PROTON-2349) and will be monitoring this closely

**Comment by timtay-microsoft** [comment]
It looks like the proton-j team finally checked in the fix for this bug [here](https://github.com/apache/qpid-proton-j/commit/7071772619d8bc596b30b01cc7816118abc9e598) so we will upgrade this library's proton-j version as soon as they put out a release with this new fix.

**Comment by timtay-microsoft** [comment]
This bug has been fixed as of device client version 2.1.1, as of service client version 2.1.2, and as of provisioning device client version 2.0.2, so I'm closing this issue
